self-core-impl/src/main/java/com/selfxdsd/core/CachingJsonResources.java

```diff
@@ -0,0 +1,241 @@
+     */
+    private final JsonStorage jsonStorage;
+
+    /**
+     * Ctor.
+     * @param delegate JsonResources delegate.
+     * @param jsonStorage JSON storage.
+     */
+    public CachingJsonResources(final JsonResources delegate,
+                                final JsonStorage jsonStorage) {
+        this.delegate = delegate;
+        this.jsonStorage = jsonStorage;
+    }
+
+    /**
+     * Ctor.
+     * @param delegate JsonResources delegate.
+     */
+    public CachingJsonResources(final JsonResources delegate) {
+        this(delegate, new JsonStorage.InMemory());
+    }
+
+
+    @Override
+    public JsonResources authenticated(final AccessToken accessToken) {
+        return new CachingJsonResources(
+            this.delegate.authenticated(accessToken),
+            this.jsonStorage
+        );
+    }
+
+    @Override
+    public Resource get(final URI uri) {
+        return this.tryGetFromCache(uri, Collections::emptyMap);
+    }
+
+    @Override
+    public Resource get(final URI uri,
+                        final Supplier<Map<String, List<String>>> headers) {
+        return this.tryGetFromCache(uri, headers);
+    }
+
+    @Override
+    public Resource post(final URI uri,
+                         final JsonValue body) {
+        return this.delegate.post(uri, body);
+    }
+
+    @Override
+    public Resource post(final URI uri,
+                         final Supplier<Map<String, List<String>>> headers,
+                         final JsonValue body) {
+        return this.delegate.post(uri, headers, body);
+    }
+
+    @Override
+    public Resource patch(final URI uri, final JsonValue body) {
+        return this.delegate.patch(uri, body);
+    }
+
+    @Override
+    public Resource put(final URI uri, final JsonValue body) {
+        return this.delegate.put(uri, body);
+    }
+
+    @Override
+    public Resource delete(final URI uri, final JsonValue body) {
+        return this.delegate.delete(uri, body);
+    }
+
+    /**
+     * Try to get Resource from json storage cache by URI, otherwise fetch
+     * from remote and then store the Resource.
+     * @param uri URI.
+     * @param headers Current Headers.
+     * @return Cached or remote Resource.
+     */
+    private Resource tryGetFromCache(
+        final URI uri,
+        final Supplier<Map<String, List<String>>> headers
+    ) {
+        final Resource resource;
+        String cachedEtag = this.jsonStorage.getEtag(uri);
+        if (cachedEtag != null) {
+            final Supplier<Map<String, List<String>>> updatedHeaders =
+                this.updateHeaders(headers, cachedEtag);
+            final Resource remoteResource = this.delegate
+                .get(uri, updatedHeaders);
+            final int status = remoteResource.statusCode();
+
+            if (status == HttpURLConnection.HTTP_NOT_MODIFIED) {
+                final String cachedBody = this.jsonStorage.getResourceBody(uri);
+                if (cachedBody != null) {
```

Comment:
@criske Is this if/else really necessary? If the status code is NOT_MODIFIED, then we should always find the body in the storage, or?

Reply:
@amihaiemil For current JsonStorage implementation, no... but if other implementations will keep etag and body in separate keys and support key eviction, we need to double check for a missing cache key.